File: src/main/java/fi/metatavu/oioi/cm/persistence/dao/ResourceDAO.java
package fi.metatavu.oioi.cm.persistence.dao;

import javax.enterprise.context.ApplicationScoped;
import javax.persistence.EntityManager;
import javax.persistence.TypedQuery;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;

import java.util.List;
import java.util.UUID;

import fi.metatavu.oioi.cm.model.ResourceType;
import fi.metatavu.oioi.cm.persistence.model.*;

/**
 * DAO class for Resource
 * 
 * @author Antti Leppä
 */
@ApplicationScoped
public class ResourceDAO extends AbstractDAO<Resource> {

  /**
   * Creates new Resource
   * 
   * @param id id
   * @param orderNumber orderNumber
   * @param data data
   * @param keycloakResorceId keycloakResorceId
   * @param name name
   * @param parent parent
   * @param slug slug
   * @param type type
   * @param creatorId creator's id
   * @param lastModifierId last modifier's id
   * @return created resource
   */
  @SuppressWarnings ("squid:S00107")
  public Resource create(UUID id, Integer orderNumber, String data, UUID keycloakResorceId, String name, Resource parent, String slug, ResourceType type, UUID creatorId, UUID lastModifierId) {
    Resource resource = new Resource();
    resource.setData(data);
    resource.setKeycloakResorceId(keycloakResorceId);
    resource.setName(name);
    resource.setOrderNumber(orderNumber);
    resource.setParent(parent);
    resource.setSlug(slug);
    resource.setType(type);
    resource.setId(id);
    resource.setCreatorId(creatorId);
    resource.setLastModifierId(lastModifierId);
    return persist(resource);
  }

  /**
   * Lists resources by parent
   * 
   * @param parent parent
   * @return List of applications
   */
  public List<Resource> listByParent(Resource parent) {
    EntityManager entityManager = getEntityManager();

    CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
    CriteriaQuery<Resource> criteria = criteriaBuilder.createQuery(Resource.class);
    Root<Resource> root = criteria.from(Resource.class);

    criteria.select(root);
    criteria.where(criteriaBuilder.equal(root.get(Resource_.parent), parent));
    criteria.orderBy(criteriaBuilder.asc(root.get(Resource_.orderNumber)));
    
    TypedQuery<Resource> query = entityManager.createQuery(criteria);
    
    return query.getResultList();
  }

  /**
   * Updates order number
   *
   * @param orderNumber order number
   * @param lastModifierId last modifier's id
   * @return updated resource
   */
  public Resource updateOrderNumber(Resource resource, Integer orderNumber, UUID lastModifierId) {
    resource.setLastModifierId(lastModifierId);
    resource.setOrderNumber(orderNumber);
    return persist(resource);
  }

  /**
   * Updates data
   *
   * @param data data
   * @param lastModifierId last modifier's id
   * @return updated resource
   */
  public Resource updateData(Resource resource, String data, UUID lastModifierId) {
    resource.setLastModifierId(lastModifierId);
    resource.setData(data);
    return persist(resource);
  }

  /**
   * Updates keycloakResorceId
   *
   * @param keycloakResorceId keycloakResorceId
   * @param lastModifierId last modifier's id
   * @return updated resource
   */
  public Resource updateKeycloakResorceId(Resource resource, UUID keycloakResorceId, UUID lastModifierId) {
    resource.setLastModifierId(lastModifierId);
    resource.setKeycloakResorceId(keycloakResorceId);
    return persist(resource);
  }

  /**
   * Updates name
   *
   * @param name name
   * @param lastModifierId last modifier's id
   * @return updated resource
   */
  public Resource updateName(Resource resource, String name, UUID lastModifierId) {
    resource.setLastModifierId(lastModifierId);
    resource.setName(name);
    return persist(resource);
  }

  /**
   * Updates parent
   *
   * @param parent parent
   * @param lastModifierId last modifier's id
   * @return updated resource
   */
  public Resource updateParent(Resource resource, Resource parent, UUID lastModifierId) {
    resource.setLastModifierId(lastModifierId);
    resource.setParent(parent);
    return persist(resource);
  }

  /**
   * Updates slug
   *
   * @param slug slug
   * @param lastModifierId last modifier's id
   * @return updated resource
   */
  public Resource updateSlug(Resource resource, String slug, UUID lastModifierId) {
    resource.setLastModifierId(lastModifierId);
    resource.setSlug(slug);
    return persist(resource);
  }

  /**
   * Updates type
   *
   * @param type type
   * @param lastModifierId last modifier's id
   * @return updated resource
   */
  public Resource updateType(Resource resource, ResourceType type, UUID lastModifierId) {
    resource.setLastModifierId(lastModifierId);
    resource.setType(type);
    return persist(resource);
  }

}

File: src/main/kotlin/fi/metatavu/oioi/cm/persistence/dao/ResourceDAO.kt
package fi.metatavu.oioi.cm.persistence.dao

import fi.metatavu.oioi.cm.model.ResourceType
import java.util.UUID
import fi.metatavu.oioi.cm.persistence.model.*
import javax.enterprise.context.ApplicationScoped

/**
 * DAO class for Resource
 *
 * @author Antti Leppä
 */
@ApplicationScoped
class ResourceDAO : AbstractDAO<Resource>() {

    /**
     * Creates new Resource
     *
     * @param id id
     * @param orderNumber orderNumber
     * @param data data
     * @param keycloakResourceId keycloakResorceId
     * @param name name
     * @param parent parent
     * @param slug slug
     * @param type type
     * @param creatorId creator's id
     * @param lastModifierId last modifier's id
     * @return created resource
     */
    fun create(
        id: UUID?,
        orderNumber: Int?,
        data: String?,
        keycloakResourceId: UUID?,
        name: String?,
        parent: Resource?,
        slug: String?,
        type: ResourceType?,
        creatorId: UUID?,
        lastModifierId: UUID?
    ): Resource {
        val resource = Resource()
        resource.data = data
        resource.keycloakResorceId = keycloakResourceId
        resource.name = name
        resource.orderNumber = orderNumber
        resource.parent = parent
        resource.slug = slug
        resource.type = type
        resource.id = id
        resource.creatorId = creatorId
        resource.lastModifierId = lastModifierId
        return persist(resource)
    }

    /**
     * Lists resources by parent
     *
     * @param parent parent
     * @return List of resources
     */
    fun listByParent(parent: Resource): List<Resource> {
        val entityManager = entityManager
        val criteriaBuilder = entityManager.criteriaBuilder
        val criteria = criteriaBuilder.createQuery(Resource::class.java)
        val root = criteria.from(Resource::class.java)
        criteria.select(root)
        criteria.where(criteriaBuilder.equal(root.get(Resource_.parent), parent))
        criteria.orderBy(criteriaBuilder.asc(root.get(Resource_.orderNumber)))
        val query = entityManager.createQuery(criteria)
        return query.resultList
    }

    /**
     * Find resource by parent and slug
     *
     * @param parent parent
     * @param slug slug
     * @return resource or null if not found
     */
    fun findByParentAndSlug(parent: Resource, slug: String): Resource? {
        val entityManager = entityManager
        val criteriaBuilder = entityManager.criteriaBuilder
        val criteria = criteriaBuilder.createQuery(Resource::class.java)
        val root = criteria.from(Resource::class.java)

        criteria.select(root)
        criteria.where(
            criteriaBuilder.and(
                criteriaBuilder.equal(root.get(Resource_.parent), parent),
                criteriaBuilder.equal(root.get(Resource_.slug), slug)
            )
        )

        return getSingleResult(entityManager.createQuery(criteria))
    }

    /**
     * Find resource by parent and name
     *
     * @param parent parent
     * @param name name
     * @return resource or null if not found
     */
    fun findByParentAndName(parent: Resource, name: String): Resource? {
        val entityManager = entityManager
        val criteriaBuilder = entityManager.criteriaBuilder
        val criteria = criteriaBuilder.createQuery(Resource::class.java)
        val root = criteria.from(Resource::class.java)

        criteria.select(root)
        criteria.where(
            criteriaBuilder.and(
                criteriaBuilder.equal(root.get(Resource_.parent), parent),
                criteriaBuilder.equal(root.get(Resource_.name), name)
            )
        )

        return getSingleResult(entityManager.createQuery(criteria))
    }

    /**
     * Updates order number
     *
     * @param orderNumber order number
     * @param lastModifierId last modifier's id
     * @return updated resource
     */
    fun updateOrderNumber(resource: Resource, orderNumber: Int?, lastModifierId: UUID?): Resource? {
        resource.lastModifierId = lastModifierId
        resource.orderNumber = orderNumber
        return persist(resource)
    }

    /**
     * Updates data
     *
     * @param data data
     * @param lastModifierId last modifier's id
     * @return updated resource
     */
    fun updateData(resource: Resource, data: String?, lastModifierId: UUID?): Resource? {
        resource.lastModifierId = lastModifierId
        resource.data = data
        return persist(resource)
    }

    /**
     * Updates keycloakResourceId
     *
     * @param keycloakResourceId keycloakResourceId
     * @param lastModifierId last modifier's id
     * @return updated resource
     */
    fun updateKeycloakResourceId(resource: Resource, keycloakResourceId: UUID?, lastModifierId: UUID?): Resource? {
        resource.lastModifierId = lastModifierId
        resource.keycloakResorceId = keycloakResourceId
        return persist(resource)
    }

    /**
     * Updates name
     *
     * @param name name
     * @param lastModifierId last modifier's id
     * @return updated resource
     */
    fun updateName(resource: Resource, name: String?, lastModifierId: UUID?): Resource? {
        resource.lastModifierId = lastModifierId
        resource.name = name
        return persist(resource)
    }

    /**
     * Updates parent
     *
     * @param parent parent
     * @param lastModifierId last modifier's id
     * @return updated resource
     */
    fun updateParent(resource: Resource, parent: Resource?, lastModifierId: UUID?): Resource? {
        resource.lastModifierId = lastModifierId
        resource.parent = parent
        return persist(resource)
    }

    /**
     * Updates slug
     *
     * @param slug slug
     * @param lastModifierId last modifier's id
     * @return updated resource
     */
    fun updateSlug(resource: Resource, slug: String?, lastModifierId: UUID?): Resource? {
        resource.lastModifierId = lastModifierId
        resource.slug = slug
        return persist(resource)
    }

    /**
     * Updates type
     *
     * @param type type
     * @param lastModifierId last modifier's id
     * @return updated resource
     */
    fun updateType(resource: Resource, type: ResourceType?, lastModifierId: UUID?): Resource? {
        resource.lastModifierId = lastModifierId
        resource.type = type
        return persist(resource)
    }
}